File: frontend/src/types/connection.ts
export type ConnectionType = 'mssql' | 'bigquery'

export interface MssqlConfig {
  host: string
  port: number
  database: string
  username: string
  password: string
  driver: string
  trust_server_certificate: boolean
  encrypt: boolean
}

export interface BigQueryConfig {
  project_id: string
  dataset: string
  credentials_json: string
}

export type ConnectionConfig = MssqlConfig | BigQueryConfig

export interface Connection {
  id: string
  name: string
  type: ConnectionType
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface ConnectionDetail extends Connection {
  config: Record<string, unknown>
}

export interface ConnectionCreate {
  name: string
  type: ConnectionType
  config: ConnectionConfig
}

export interface ConnectionUpdate {
  name?: string
  config?: ConnectionConfig
  is_active?: boolean
}

export interface ConnectionTestResult {
  success: boolean
  message: string
  details?: Record<string, unknown>
}

export interface SchemaInfo {
  name: string
}

export interface TableInfo {
  name: string
  schema_name: string
  row_count: number | null
}

export interface ColumnInfo {
  name: string
  data_type: string
  nullable: boolean
  max_length: number | null
  is_primary_key: boolean
}

File: backend/app/schemas/connection.py
from __future__ import annotations

from datetime import datetime
from typing import Optional, Union

from pydantic import BaseModel, Field


class MssqlConfig(BaseModel):
    host: str = Field(..., description="Sunucu adresi")
    port: int = Field(1433, description="Port numarası")
    database: str = Field(..., description="Veritabanı adı")
    username: str = Field(..., description="Kullanıcı adı")
    password: str = Field(..., description="Şifre")
    driver: str = Field("ODBC Driver 18 for SQL Server", description="ODBC sürücüsü")
    trust_server_certificate: bool = Field(True, description="Sunucu sertifikasına güven")
    encrypt: bool = Field(False, description="Bağlantıyı şifrele")


class BigQueryConfig(BaseModel):
    project_id: str = Field(..., description="Google Cloud proje ID")
    dataset: str = Field("", description="Varsayılan dataset")
    credentials_json: str = Field(..., description="Service account JSON key içeriği")


class ConnectionCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255, description="Bağlantı adı")
    type: str = Field(..., pattern="^(mssql|bigquery)$", description="Bağlantı tipi")
    config: Union[MssqlConfig, BigQueryConfig]


class ConnectionUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    config: Optional[Union[MssqlConfig, BigQueryConfig]] = None
    is_active: Optional[bool] = None


class ConnectionResponse(BaseModel):
    id: str
    name: str
    type: str
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ConnectionDetail(ConnectionResponse):
    config: dict  # şifresi çözülmüş config (password maskelenmiş)


class ConnectionTestResult(BaseModel):
    success: bool
    message: str
    details: Optional[dict] = None


class SchemaInfo(BaseModel):
    name: str


class TableInfo(BaseModel):
    name: str
    schema_name: str
    row_count: Optional[int] = None


class ColumnInfo(BaseModel):
    name: str
    data_type: str
    nullable: bool = True
    max_length: Optional[int] = None
    is_primary_key: bool = False
